File: ios/SsdidWallet/Domain/Did/DidKeyResolver.swift
import Foundation

final class DidKeyResolver: DidResolver {
    func resolve(did: String) async throws -> DidDocument {
        guard did.hasPrefix("did:key:") else {
            throw DidResolutionError.unsupportedMethod(did)
        }
        let methodSpecificId = String(did.dropFirst("did:key:".count))
        guard methodSpecificId.hasPrefix("z") else {
            throw DidResolutionError.invalidMultibase
        }
        guard let decoded = Base58.decode(String(methodSpecificId.dropFirst())) else {
            throw DidResolutionError.invalidMultibase
        }
        let (codec, _) = try Multicodec.decode(decoded)

        let vmType: String
        switch codec {
        case Multicodec.ed25519Pub: vmType = "Ed25519VerificationKey2020"
        case Multicodec.p256Pub: vmType = "EcdsaSecp256r1VerificationKey2019"
        case Multicodec.p384Pub: vmType = "EcdsaSecp384VerificationKey2019"
        default: throw DidResolutionError.unsupportedCodec(codec)
        }

        let keyId = "\(did)#\(methodSpecificId)"
        let vm = VerificationMethod(
            id: keyId, type: vmType, controller: did, publicKeyMultibase: methodSpecificId
        )
        return DidDocument(
            id: did, controller: did, verificationMethod: [vm],
            authentication: [keyId], assertionMethod: [keyId], capabilityInvocation: [keyId]
        )
    }
}

File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/did/DidKeyResolver.kt
package my.ssdid.sdk.domain.did

import my.ssdid.sdk.domain.model.DidDocument
import my.ssdid.sdk.domain.model.VerificationMethod

class DidKeyResolver : DidResolver {
    override suspend fun resolve(did: String): Result<DidDocument> = runCatching {
        require(did.startsWith("did:key:")) { "Not a did:key: $did" }

        val methodSpecificId = did.removePrefix("did:key:")
        require(methodSpecificId.startsWith("z")) { "Expected multibase 'z' (base58btc) prefix" }

        val decoded = Base58.decode(methodSpecificId.substring(1))
        val (codec, _) = Multicodec.decode(decoded)

        val vmType = when (codec) {
            Multicodec.ED25519_PUB -> "Ed25519VerificationKey2020"
            Multicodec.P256_PUB -> "EcdsaSecp256r1VerificationKey2019"
            Multicodec.P384_PUB -> "EcdsaSecp384VerificationKey2019"
            else -> throw IllegalArgumentException("Unsupported multicodec: 0x${codec.toString(16)}")
        }

        val keyId = "$did#$methodSpecificId"

        val vm = VerificationMethod(
            id = keyId,
            type = vmType,
            controller = did,
            publicKeyMultibase = methodSpecificId
        )

        DidDocument(
            id = did,
            controller = did,
            verificationMethod = listOf(vm),
            authentication = listOf(keyId),
            assertionMethod = listOf(keyId),
            capabilityInvocation = listOf(keyId)
        )
    }
}
